### DISCUSSION TITLE
Proper way to use Enum with Pydantic & FastAPI

### DISCUSSION THREAD

**Discussion by brymon68** [pydantic/pydantic#4967] [Question] [answered]
Hello, new to Pydantic & FastAPI - 

I have an Enum: 
```
class ColorEnum(str, Enum):
    orange = "orange"
    red = "red"
    green = "green"
```

Which I use to define a Model such as: 

```
class Palette(BaseModel):
    colors: List[ColorEnum]
    # other attributes

    @validator("colors")
    def colors_are_valid(cls, colors):
        color_values = set(item.value for item in ColorEnum)
        if not all(map(lambda x: x in color_values, colors)):
            raise ValueError("Colors must be red, green or orange")
        return colors
```

And I have a request hander that looks like: 

```
@router.post("/create")
def create_palette(request: Request, item: Palette):
 
    colors = item.colors
    # do other stuff
```

When I submit a POST to my handler, that looks like: 
```
{
   "colors": ["red","green"]
   // other stuff
}
```
I receive the following error:
```
Status Code: 422 -1 validation error for Request body -> integrations -> 
0 value is not a valid enumeration member; permitted: 'orange', 'red', 'green'
(type=type_error.enum; enum_values=[<ColorEnum.orange: 'orange'>, <ColorEnum.red: 'red'>])
````

How should I be better handling this to map whatever strings (colors in this case) to the correct Enum values to be validated? 

Thank you in advance

**Comment by brymon68** [comment, marked as answer]
This can be resolved. I ended up moving this changing my Palette color attribute to be that of type string, but still handle the type checking at the validator which seems to work fine.

**Comment by nymous** [comment]
You can have no validator at all :smiley: 

This works fine and converts the JSON strings into the corresponding enum members:
```py
from enum import Enum

from fastapi import FastAPI
from pydantic import BaseModel

class ColorEnum(str, Enum):
    orange = "orange"
    red = "red"
    green = "green"

class Palette(BaseModel):
    colors: list[ColorEnum]

app = FastAPI()

@app.post("/create")
def create_palette(item: Palette):
    print(item)
    print(item.colors)
```

With the body you showed, this prints
```
colors=[<ColorEnum.red: 'red'>, <ColorEnum.green: 'green'>]
[<ColorEnum.red: 'red'>, <ColorEnum.green: 'green'>]
```

It also handles invalid enum members, for example if you post `{"colors": ["blue"]}` you will get the following response:
```json
{
  "detail": [
    {
      "loc": [
        "body",
        "colors",
        0
      ],
      "msg": "value is not a valid enumeration member; permitted: 'orange', 'red', 'green'",
      "type": "type_error.enum",
      "ctx": {
        "enum_values": [
          "orange",
          "red",
          "green"
        ]
      }
    }
  ]
}
```

**Reply by TijmenSimons** [reply]
This is what I have, and works just like you explained.

But now I'm having trouble with the objects I'm getting in my request function as I'm trying to put it directly into my database.
Do I have to convert it from the Enum Value to a normal string myself or is there a nice way to fix it for every Enum I have?

My model column:
```python
status: Mapped[SwipeSessionEnum] = mapped_column(
        default=SwipeSessionEnum.IN_PROGRESS
)
```
My schema:
```python
status: SwipeSessionEnum = SwipeSessionEnum.IN_PROGRESS
```

My enum:
```python
class SwipeSessionEnum(str, BaseEnum):
    PAUSED = "Gepauzeerd"
    CANCELLED = "Gestopt"
    IN_PROGRESS = "Is bezig"
    COMPLETED = "Voltooid"
```

My error:
```
sqlalchemy.exc.DBAPIError: (sqlalchemy.dialects.postgresql.asyncpg.Error) <class 'asyncpg.exceptions.InvalidTextRepresentationError'>: invalid input value for enum swipesessionenum: "IN_PROGRESS"
[SQL: INSERT INTO swipe_session (status, user_id) VALUES ($1::swipesessionenum, $2::INTEGER) RETURNING swipe_session.id]
[parameters: ('IN_PROGRESS', 1)]
```

It seems to be putting in the enum object instead of just its value?

EDIT:

I fixed it by adding the `type_annotation_map` to the Base object for the models: https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-python-enum-or-pep-586-literal-types-in-the-type-map

**Comment by Kilo59** [comment]
@brymon68 Have you tried using the `use_enum_values` model config option?

>**use_enum_values**:
> whether to populate models with the value property of enums, rather than the raw enum.

https://docs.pydantic.dev/usage/model_config/


```python
class Palette(BaseModel):
    colors: List[ColorEnum]

   class Config:
       use_enum_values = True
```

You could also use a `Literal`

```python
class Palette(BaseModel):
    colors: List[Literal["orange", "red", "green"]]
```
